pageserver/src/tenant.rs

```diff
@@ -978,81 +985,88 @@ impl Tenant {
         let mut totals: GcResult = Default::default();
         let now = Instant::now();
 
-        // grab mutex to prevent new timelines from being created here.
-        let gc_cs = self.gc_cs.lock().unwrap();
+        let gc_timelines = {
+            // grab mutex to prevent new timelines from being created here.
+            let gc_cs = self.gc_cs.lock().unwrap();
 
-        let timelines = self.timelines.lock().unwrap();
+            let timelines = self.timelines.lock().unwrap();
 
-        // Scan all timelines. For each timeline, remember the timeline ID and
-        // the branch point where it was created.
-        let mut all_branchpoints: BTreeSet<(TimelineId, Lsn)> = BTreeSet::new();
-        let timeline_ids = {
-            if let Some(target_timeline_id) = target_timeline_id.as_ref() {
-                if timelines.get(target_timeline_id).is_none() {
-                    bail!("gc target timeline does not exist")
-                }
-            };
-
-            timelines
-                .iter()
-                .filter(|(_, timeline)| timeline.is_active())
-                .map(|(timeline_id, timeline_entry)| {
-                    // This is unresolved question for now, how to do gc in presence of remote timelines
-                    // especially when this is combined with branching.
-                    // Somewhat related: https://github.com/neondatabase/neon/issues/999
-                    if let Some(ancestor_timeline_id) = &timeline_entry.get_ancestor_timeline_id() {
-                        // If target_timeline is specified, we only need to know branchpoints of its children
-                        if let Some(timeline_id) = target_timeline_id {
-                            if ancestor_timeline_id == &timeline_id {
+            // Scan all timelines. For each timeline, remember the timeline ID and
+            // the branch point where it was created.
+            let mut all_branchpoints: BTreeSet<(TimelineId, Lsn)> = BTreeSet::new();
+            let timeline_ids = {
+                if let Some(target_timeline_id) = target_timeline_id.as_ref() {
+                    if timelines.get(target_timeline_id).is_none() {
+                        bail!("gc target timeline does not exist")
+                    }
+                };
+
+                timelines
+                    .iter()
+                    // TODO here was the filter to take only active timelines. Is it actually correct?
```

Comment:
It is out of scope for the PR but just a change I've noticed. Here in main we filter inactive timelines which I think is wrong. Gc needs to know about all branch points, not only ones for timelines that are active at the moment of gc. If timeline is inactive then we wont know about branch point and I believe in this case gc can delete data that is needed by child timeline.

Reply:
Agreed

Reply:
Fixed in separate PR #2726